README.md
# minimalcss

![Build status](https://github.com/peterbe/minimalcss/workflows/Node.js%20CI/badge.svg)
[![NPM version](https://img.shields.io/npm/v/minimalcss.svg)](https://www.npmjs.com/package/minimalcss)
[![styled with prettier](https://img.shields.io/badge/styled_with-prettier-ff69b4.svg)](#badge)
[![Renovate enabled](https://img.shields.io/badge/renovate-enabled-brightgreen.svg)](https://renovateapp.com/)

A Node library to extract the minimal CSS used in a set of URLs with puppeteer.
Used to find what minimal CSS is needed to render on first load, even with
`document.onload` executed.

This minimal CSS is also known as [critical path CSS](https://developers.google.com/web/fundamentals/performance/critical-rendering-path/analyzing-crp)
and ultimately a web performance technique to make web pages load faster
at initial load.

## What does it do

You supply a list of URLs that it opens (one at a time) and for each page
it downloads all external CSS files (e.g.
`<link rel="stylesheet" href="bootstrap.min.css">`) and uses the DOM and
`document.querySelector` to investigate which selectors, in the CSS, are
actually in the DOM. That minimal payload of CSS is all you need to load
the URLs styled without having to make it block on CSS.

Under the hood it relies on the excellent
[puppeteer library](https://github.com/GoogleChrome/puppeteer) which uses
the Headless Chome Node API. This means it runs (at the time of writing)
Chrome 62 and this library is maintained by the Google Chrome team.

The CSS to analyze (and hopefully minimize) is downloaded automatically just
like a browser opens and downloads CSS as mentioned in the DOM as `<link>`
tags.

The CSS is parsed by [CSSTree](https://github.com/csstree/csstree) and the
minification and compression is done with [CSSO](https://github.com/css/csso).
An AST of each CSS payload is sent into the Headless Chrome page evaluation
together with a callback that compares with the DOM and then each minimal CSS
payload is concatenated into one big string which then CSSO compresses into
one "merged" and minified CSS payload.

## Usage

Install:

```bash
yarn add minimalcss --dev
```

You can install it globally if you like:

```bash
yarn global add minimalcss
```

```bash
npm install [--save-dev|--global] minimalcss
```

Now you can run it:

```bash
./node_modules/.bin/minimalcss https://example.com/ https://example.com/aboutus > minimal.min.css
```

## Prior art

`minimalcss` isn't the first library to perform this task. What's unique and
special about `minimalcss` is that it uses the Chrome Headless browser.

* [penthouse](https://github.com/pocketjoso/penthouse) -
  uses [puppeteer](https://github.com/GoogleChrome/puppeteer) (since version 1.0) and [CSSTree](https://github.com/csstree/csstree).
  Supports only 1 URL at a time and can't you have to first save the CSS files
  it should process.

* [critical](https://github.com/addyosmani/critical) - uses `penthouse`
  (see above) with its "flaws" meaning you can only do 1 URL (or HTML string)
  and you have to prepare the CSS files too.

* [UnCSS](https://github.com/giakki/uncss) - uses [jsdom](https://github.com/tmpvar/jsdom)
  to render and execute JavaScript. Supports supplying multiple URLs but still
  requires to manually supply the CSS files to process.

* [mincss](https://github.com/peterbe/mincss) - Python project that uses
  [lxml.html](http://lxml.de/lxmlhtml.html) to analyze the HTML statically
  (by doing a `GET` of the URL as if done by a server). I.e.
  it can't load the HTML as a real browser would and thus does not support a
  DOM with possible JavaScript mutations on load.
  It can optionally use `PhantomJS` to extract the HTML.

## Killer features

* You don't need to specify where the CSS is. It gets downloaded and parsed
  automatically.

* It uses [puppeteer](https://github.com/GoogleChrome/puppeteer) and
  [CSSTree](https://github.com/csstree/csstree) which are both high quality
  projects that are solid and well tested.

* The CSS selectors downloaded is compared to the DOM before _and_ after
  JavaScript code has changed the DOM. That means you can extract the
  critical CSS needed to display properly before the JavaScript has kicked in.

* Ability to analyze the remaining CSS selectors to see which keyframe
  animations that they use and use this to delete keyframe definitions
  that are no longer needed.

* You can specify a [viewport](https://github.com/GoogleChrome/puppeteer/blob/v1.0.0/docs/api.md#pagesetviewportviewport),
  which might cause the page to render slightly different. It does not
  create the minimal CSS _only_ on DOM that is visible though.

* If the CSS contains `@font-face { ... }` rules whose name is never
  used in any remaining CSS selector, the whole `font-face` block is removed.

## Help needed

Let's make this a thriving community project!

Help needed with features, tooling, and much testing in real web performance
optimization work.

## API

```javascript
const minimalcss = require('minimalcss');
```

### Get version `minimalcss.version`

Just prints out the current version.

### Run a minimization `minimalcss.run(options)`

Returns a promise. The promise returns an object containing, amongst
other things, the minified minimal CSS as a string.
For example:

```javascript
minimalcss
  .minimize({ url: 'https://example.com/' })
  .then(result => {
    console.log('OUTPUT', result.finalCss.length, result.finalCss);
  })
  .catch(error => {
    console.error(`Failed the minimize CSS: ${error}`);
  });
```

That `result` object that is returned by the `minimize` function contains:

* `finalCss` - the minified minimal CSS as a string.
* `stylesheetContents` - an object of stylesheet URLs as keys and their
  content as text.

Optionally, you can supply a list of URLs like this:

```javascript
minimalcss
  .minimize({ urls: ['https://example.com/page1', 'https://example.com/page2'] })
  ...
```

and `minimalcss` will try to merge the minimal critical CSS across all pages.
But we aware that this can be "dangerous" because of the inherit order of CSS.

## API Options

Calling `minimalcss.run(options)` takes an object whose only mandatory
key is `urls` or `url`. Other optional options are:

* `debug` - all console logging during page rendering are included in the
  stdout. Also, any malformed selector that cause errors in `document.querySelector`
  will be raised as new errors.
* `skippable` - function which takes
  [request](https://pptr.dev/#?product=Puppeteer&show=api-class-httprequest)
  as an argument and returns boolean. If it returns true then given request
  will be aborted (skipped). Can be used to block requests to Google Analytics
  etc.
* `loadimages` - If set to `true`, images will actually load.
* `withoutjavascript` - If set to `false` it will _skip_ loading the page first
  without JavaScript. By default `minimalcss` will evaluate the DOM as plain as
  can be, and then with JavaScript enabled _and_ waiting for network activity
  to be idle.
* `disableJavaScript` - By default JavaScript is enabled. If set to `true` it will ignore `withoutjavascript` option and loading the page only one time without JavaScript.
* `browser` - Instance of a [Browser](https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#class-browser), which will be used instead of launching another one.
* `userAgent` - specific user agent to use (string)
* `viewport` - viewport object as specified in [page.setViewport](https://github.com/GoogleChrome/puppeteer/blob/v1.0.0/docs/api.md#pagesetviewportviewport)
* `puppeteerArgs` - Args sent to puppeteer when launching. [List
  of strings for headless Chrome](https://peter.sh/experiments/chromium-command-line-switches/).
* `cssoOptions` - CSSO compress function [options](https://github.com/css/csso#compressast-options)
* `timeout` - Maximum navigation time in milliseconds, defaults to 30 seconds, pass 0 to disable timeout.
* `ignoreCSSErrors` - By default, any CSS parsing error throws an error in `minimalcss`. If you know it's safe to ignore (for example, third-party CSS resources), set this to `true`.

CHANGELOG.md
# 0.11.3

* Cope with selectors with a `\` character before the `:`
  [pull#426](https://github.com/peterbe/minimalcss/pull/426)

# 0.11.2

* Add a `.npmignore` to avoid shipping tests and backup files.

# 0.11.1

* Upgrade sub-dependency `find-my-way`
  [pull#407](https://github.com/peterbe/minimalcss/pull/407)

# 0.11.0

* The main parameter `urls` can no be a string called just `url`. You just
  need to either use `urls` (array) or `url` (string).

* New `ignoreRequestErrors` option to ignore HTTP responses that are `>=400`.
  [pull#365](https://github.com/peterbe/minimalcss/pull/365)
  Thanks @Fgruntjes

* (Chore) Switching from TravisCI to GitHub Actions

# 0.10.0

* Upgrade dependency on `puppeteer` to `^2.0.0`

# 0.9.0

* Ability to pass a list of whitelist selector regexes
  [pull#344](https://github.com/peterbe/minimalcss/pull/344)
  Thanks @AlexDubok

* Upgrade dependency on `css-tree` to `1.0.0-alpha.37`

* Upgrade dependency on `csso` to `~4.0.2`

# 0.8.3

* You can now render exclusively with JavaScript disabled.
  [pull#312](https://github.com/peterbe/minimalcss/pull/312)
  Thanks @VladislavTkachuk

# 0.8.2

* Big optimization. By looking up the "parent CSS selector" pre-emptively
  it can quickly discard repeated "child CSS selectors".
  [pull#296](https://github.com/peterbe/minimalcss/pull/296)

# 0.8.1

* Ignore `<link rel="preload">` and `<link rel="prefetch">` tags whose
  `href` value was a `.css` extension.
  [pull#275](https://github.com/peterbe/minimalcss/pull/275)
  Thanks @nicolas-t

# 0.8.0

* By default, Service Workers are disabled. This new feature requires an
  upgrade of puppeteer to `^1.8.0`.
  [pull#265](https://github.com/peterbe/minimalcss/pull/265)

* Ability to also extract the CSS of all `style` tags with the `styletags`
  option.
  [pull#260](https://github.com/peterbe/minimalcss/pull/260)
  Thanks @jc275

# 0.7.10

* Not crash if the CSS contains multiple semicolons which crashes
  csso (https://github.com/css/csso/issues/378)
  [pull#259](https://github.com/peterbe/minimalcss/pull/259)
  Thanks @jc275

* Stylesheet `link` tags whose `href` URL contains a `#fragment-example`
  would cause an error because puppeteer doesn't include it in the
  `response.url()`.
  [pull#255](https://github.com/peterbe/minimalcss/pull/255)
  Thanks @jc275

# 0.7.9

* New option `ignoreJSErrors` to ignore possible JavaScript errors.
  [pull#253](https://github.com/peterbe/minimalcss/pull/253)
  Thanks @jc275

# 0.7.8

* New option `ignoreCSSErrors` to ignore possible CSS parsing errors.
  [pull#249](https://github.com/peterbe/minimalcss/pull/249)
  Thanks @stereobooster

# 0.7.7

* Throw explicit errors on invalid CSS
  [pull#237](https://github.com/peterbe/minimalcss/pull/237)
  Thanks @harrygreen

# 0.7.6

* List what timed out. Useful for debugging which resources failed.
  [pull#199](https://github.com/peterbe/minimalcss/pull/199)
  Thanks @stereobooster

* Upgrade to puppeteer 1.4.0
  [pull#214](https://github.com/peterbe/minimalcss/pull/214)

# 0.7.5

* Ability to pass an object of options to `csso.compress()`
  [pull#167](https://github.com/peterbe/minimalcss/pull/167)
  Thanks @usebaz

# 0.7.4

* Fix for logic of using the `--withoutjavascript` argument.
  [pull#163](https://github.com/peterbe/minimalcss/pull/163)
  Thanks @stereobooster

* Upgrade `puppeteer` dependency to version 1.2.

* Fix for `304 Not Modified` responses that actually don't redirect.
  [pull#165](https://github.com/peterbe/minimalcss/pull/165)
  Thanks @stereobooster

# 0.7.3

* Fix for pages that uses `data:text/css ...` as the `href` i `<link>` tags.
  [pull#159](https://github.com/peterbe/minimalcss/pull/159)

# 0.7.2

* Data URIs in external stylesheets lost the `data:` part.
  [pull#153](https://github.com/peterbe/minimalcss/pull/153)
  Thanks @stereobooster and @phiresky for reporting.

# 0.7.1

* Any query strings in URLs in CSS is now preserved.
  [pull#148](https://github.com/peterbe/minimalcss/pull/148) Thanks @usebaz

# 0.7.0


src/run.js
'use strict';

const puppeteer = require('puppeteer');
// @ts-ignore
const csso = require('csso');
// @ts-ignore
const csstree = require('css-tree');
const cheerio = require('cheerio');
const utils = require('./utils');
const { createTracker } = require('./tracker');
const url = require('url');

const isOk = (response) => response.ok() || response.status() === 304;

/**
 * Take in a csstree AST, mutate it and return a csstree AST.
 * The mutation is about:
 *
 *   1) Remove all keyframe declarations that are *not* mentioned
 *      by animation name.
 *   2) Remove all font-face declarations that are *not* mentioned
 *      as the font-family.
 *
 * The gist of the function is that it walks the AST, populates sets
 * that track the names of all animations and font families. Then,
 * it converts the AST to a plain object, which it mutates by filtering
 * the 'children' object.
 * Lastly it uses the csstree's fromPlainObject to return the plain
 * object back as an AST.
 * @param {Object} ast
 * @return Object
 */
const postProcessOptimize = (ast) => {
  // First walk the AST to know which animations are ever mentioned
  // by the remaining rules.
  const activeAnimationNames = new Set(
    csstree.lexer
      .findAllFragments(ast, 'Type', 'keyframes-name')
      .map((entry) => csstree.generate(entry.nodes.first()))
  );

  // This is the function we use to filter @keyframes atrules out,
  // if its name is not actively used.
  // It also filters out all `@media print` atrules.
  csstree.walk(ast, {
    visit: 'Atrule',
    enter: (node, item, list) => {
      const basename = csstree.keyword(node.name).basename;
      if (basename === 'keyframes') {
        if (!activeAnimationNames.has(csstree.generate(node.prelude))) {
          list.remove(item);
        }
      } else if (basename === 'media') {
        if (csstree.generate(node.prelude) === 'print') {
          list.remove(item);
        }
      }
    },
  });

  // Now figure out what font-families are at all used in the AST.
  const activeFontFamilyNames = new Set();

  // Walk into every font-family rule and inspect if we uses its declarations
  csstree.walk(ast, {
    visit: 'Atrule',
    enter: (atrule, atruleItem, atruleList) => {
      if (csstree.keyword(atrule.name).basename === 'font-face') {
        // We're inside a font-face rule! Let's dig deeper.
        csstree.walk(atrule, {
          visit: 'Declaration',
          enter: (declaration) => {
            if (csstree.property(declaration.property).name === 'font-family') {
              const name = utils.unquoteString(
                csstree.generate(declaration.value)
              );
              // was this @font-face used?
              if (!activeFontFamilyNames.has(name)) {
                atruleList.remove(atruleItem);
              }
            }
          },
        });
      }
    },
  });
};

const processStylesheet = ({
  text,
  pageUrl,
  responseUrl,
  stylesheetAsts,
  stylesheetContents,
}) => {
  const ast = csstree.parse(text);
  csstree.walk(ast, (node) => {
    if (node.type !== 'Url') return;
    const value = node.value;
    let path = value.value;
    if (value.type !== 'Raw') {
      path = path.substr(1, path.length - 2);
    }
    const sameHost = url.parse(responseUrl).host === url.parse(pageUrl).host;
    if (/^https?:\/\/|^\/\/|^data:/i.test(path)) {
      // do nothing
    } else if (/^\//.test(path) && sameHost) {
      // do nothing
    } else {
      const resolved = new url.URL(path, responseUrl);
      if (sameHost) {
        path = resolved.pathname + resolved.search;
      } else {
        path = resolved.href;
      }
      if (value.type !== 'Raw') {
        value.value = `"${path}"`;
      } else {
        value.value = path;
      }
    }
  });
  stylesheetAsts[responseUrl] = ast;
  stylesheetContents[responseUrl] = text;
};

const processPage = ({
  page,
  htmlContent,
  headers,
  options,
  pageUrl,
  stylesheetAsts,
  stylesheetContents,
  doms,
  allHrefs,
  redirectResponses,
  skippedUrls,
}) =>
  new Promise(async (resolve, reject) => {
    // If anything goes wrong, for example a `pageerror` event or
    // a bad download request (e.g. !response.ok), then remember that
    // we have fulfilled the promise and don't want to call `reject` or `resolve`
    // a second time.
    let fulfilledPromise = false;

    const tracker = createTracker(page);
    const safeReject = (error) => {
      if (fulfilledPromise) return;
      fulfilledPromise = true;
      if (error.message.startsWith('Navigation timeout')) {
        const urls = tracker.urls();
        if (urls.length > 1) {
          error.message += `\nTracked URLs that have not finished: ${urls.join(
            ', '
          )}`;
        } else if (urls.length > 0) {
          error.message += `\nFor ${urls[0]}`;
        }
      }
      tracker.dispose();
      reject(error);
    };

    const debug = options.debug || false;
    const loadimages = options.loadimages || false;
    const styletags = options.styletags || false;
    const withoutjavascript =
      options.withoutjavascript === undefined
        ? true
        : !!options.withoutjavascript;
    const disableJavaScript = !!options.disableJavaScript;

    try {
      if (options.userAgent) {
        await page.setUserAgent(options.userAgent);
      }

      if (options.viewport) {
        await page.setViewport(options.viewport);
      }

      if (options.timeout !== undefined) {
        page.setDefaultNavigationTimeout(options.timeout);
      }

      const waitUntil = options.pageWaitUntil || 'networkidle0';

      // A must or else you can't do console.log from within page.evaluate()
      page.on('console', (msg) => {
        if (debug) {
          for (let i = 0; i < msg.args.length; ++i) {
            console.log(`${i}: ${msg.args[i]}`);
          }
        }
      });

      await page.setRequestInterception(true);
      page.on('request', async (request) => {
        const resourceType = request.resourceType();
        const requestUrl = request.url();
        
        if (/data:image\//.test(requestUrl)) {
          // don't need to download those
          request.abort();
        } else if (!loadimages && resourceType === 'image') {
          request.abort();
        } else if (resourceType === 'font') {
          request.abort();
        } else if (stylesheetAsts[requestUrl]) {
          // no point downloading this again
          request.abort();
        } else if (options.skippable && options.skippable(request)) {
          // If the URL of the request that got skipped is a CSS file
          // not having it in stylesheetAsts is going to cause a
          // problem later when we loop through all <link ref="stylesheet">
          // tags.
          // So put in an empty (but not falsy!) object for this URL.
          skippedUrls.add(requestUrl);
          request.abort();
        } else {
          const requestHeaders = {...request.headers(), ...headers} ;
          
          request.continue(requestHeaders);
        }
      });

      // To build up a map of all downloaded CSS
      page.on('response', (response) => {
        const responseUrl = response.url();
        const resourceType = response.request().resourceType();
        if (response.status() >= 400) {
          if (options.ignoreRequestErrors) {
            skippedUrls.add(responseUrl);
          } else {
            return safeReject(
              new Error(`${response.status()} on ${responseUrl}`)
            );
          }
        } else if (response.status() >= 300 && response.status() !== 304) {
          // If the 'Location' header points to a relative URL,
          // convert it to an absolute URL.
          // If it already was an absolute URL, it stays like that.
          const redirectsTo = new url.URL(
            response.headers().location,
            responseUrl
          ).toString();
          redirectResponses[responseUrl] = redirectsTo;
        } else if (resourceType === 'stylesheet') {
          // page has been closed already
          if (page.isClosed()) {
            debug && console.log(`Page ${pageUrl} is closed`);
            return;
          }

          response.text()
            .then((text) => {
              processStylesheet({
                  text,
                  pageUrl,
                  responseUrl,
                  stylesheetAsts,
                  stylesheetContents,
                }
              )
            })
            .catch((reason) => debug && console.log(`Cannot get text() from the response: ${ reason }`));
        }
      });

      page.on('pageerror', (error) => {
        if (!options.ignoreJSErrors) {
          safeReject(error);
        }
      });

      let response;

      if (disableJavaScript) {
        await page.setJavaScriptEnabled(false);
      } else if (withoutjavascript) {
        // First, go to the page with JavaScript disabled.
        await page.setJavaScriptEnabled(false);

        if (htmlContent) {
          response = await page.setContent(htmlContent, { waitUntil: waitUntil });
        } else {
          response = await page.goto(pageUrl, { waitUntil: waitUntil });
          if (!isOk(response)) {
            return safeReject(
              new Error(`${response.status()} on ${pageUrl} (second time)`)
            );
          }
        }
        const htmlVanilla = await page.content();
        doms.push(cheerio.load(htmlVanilla));
        await page.setJavaScriptEnabled(true);
      }

      // There is another use case *between* the pure DOM (without any
      // javascript) and after the network is idle.
      // For example, any CSSOM that is *made by javascript* but goes away
      // after all XHR is finished loading.
      // What we can do is do that lookup here using...
      //    response = await page.goto(pageUrl, {
      //        waitUntil: ['domcontentloaded']
      //    })
      // And add that to the list of DOMs.
      // This will slow down the whole processing marginally.

      // Second, goto the page and evaluate it with JavaScript.
      // The 'waitUntil' option determines how long we wait for all
      // possible assets to load.

      if (htmlContent) {
        response = await page.setContent(htmlContent, { waitUntil: waitUntil });
      } else {
        response = await page.goto(pageUrl, { waitUntil: waitUntil });
        if (!isOk(response)) {
          return safeReject(
            new Error(`${response.status()} on ${pageUrl} (second time)`)
          );
        }
      }

      const evalWithJavascript = await page.evaluate(() => {
        const html = document.documentElement.outerHTML;
        // The reason for NOT using a Set here is that that might not be
        // supported in ES5.
        const hrefs = [];
        const styles = [];
        const isCssStyleTag = (elem) =>
          elem.tagName === 'STYLE' &&
          (!elem.type || elem.type.toLowerCase() === 'text/css');

        const isStylesheetLink = (elem) =>
          elem.tagName === 'LINK' &&
          elem.href &&
          elem.rel.toLowerCase() === 'stylesheet' &&
          !elem.href.toLowerCase().startsWith('data:') &&
          !elem.href.toLowerCase().startsWith('blob:') &&
          elem.media.toLowerCase() !== 'print';

        // #fragments are omitted from puppeteer's response.url(), so
        // we need to strip them from stylesheet links, otherwise the
        // hrefs won't always match when we check for missing ASTs.
        const defragment = (href) => href.split('#')[0];
        const pageUrl = defragment(window.location.href);
        // Create a unique identifier for each style tag by appending
        // an xpath-like fragment to the page URL.  This allows us to
        // preserve the relative ordering of external stylesheets and
        // inline style tags.
        const styleTagUri = () => `${pageUrl}#style[${styles.length}]`;
        // Loop over all 'link' and 'style' elements in the document,
        // in order of appearance. For each element, collect the URI
        // of all the ones we're going to assess. For style elements,
        // also extract each tag's content.
        Array.from(document.querySelectorAll('link, style')).forEach((elem) => {
          if (isStylesheetLink(elem)) {
            const href = defragment(elem.href);
            hrefs.push(href);
          } else if (isCssStyleTag(elem)) {
            const href = styleTagUri();
            const text = elem.innerHTML;
            styles.push({ href, text });
            hrefs.push(href);
          }
        });
        return { html, hrefs, styles };
      });

      const htmlWithJavascript = evalWithJavascript.html;
      doms.push(cheerio.load(htmlWithJavascript));

      if (styletags) {
        // Parse each style tag as if it were an external stylesheet.
        evalWithJavascript.styles.forEach(({ href, text }) => {
          processStylesheet({
            text,
            pageUrl,
            stylesheetAsts,
            stylesheetContents,
            responseUrl: href,
          });
        });
      } else {
        // Remove each style tag URI from the list of hrefs.
        evalWithJavascript.styles.forEach(({ href }) => {
          evalWithJavascript.hrefs.splice(
            evalWithJavascript.hrefs.indexOf(href),
            1
          );
        });
      }

      evalWithJavascript.hrefs.forEach((href) => {
        // The order of allHrefs is important! That's what browsers do.
        // But we can't blindly using allHrefs.push() because the href
        // *might* already have been encountered. If it has been encountered
        // before, remove it and add it to the end of the array.
        if (allHrefs.includes(href)) {
          allHrefs.splice(allHrefs.indexOf(href), 1);
        }
        allHrefs.push(href);
      });

      if (!fulfilledPromise) {
        tracker.dispose();
        resolve();
      }
    } catch (e) {
      return safeReject(e);
    }
  });

/**
 *
 * @param {{ urls: Array<string>, htmlContent: string, url: string, debug: boolean, loadimages: boolean, skippable: function, browser: any, userAgent: string, withoutjavascript: boolean, viewport: any, puppeteerArgs: Array<string>, cssoOptions: Object, ignoreCSSErrors?: boolean, ignoreJSErrors?: boolean, styletags?: boolean, enableServiceWorkers?: boolean, disableJavaScript?: boolean, whitelist?: Array<string>, ignoreRequestErrors?: boolean, headers?: Object }} options
 * @return Promise<{ finalCss: string, stylesheetContents: { [key: string]: string }, doms: Array<object> }>
 */
const minimalcss = async (options) => {
  const { urls = [], url = '' } = options;
  if (url) {
    urls.push(url);
  }
  if (!urls.length) {
    throw new Error(
      "no URLs supplied (hint: use 'urls' (array) or 'url' (string)"
    );
  }
  const htmlContent = options.htmlContent || null;
  const headers = options.headers || {};
  const debug = options.debug || false;
  const cssoOptions = options.cssoOptions || {};
  const enableServiceWorkers = options.enableServiceWorkers || false;
  const puppeteerArgs = options.puppeteerArgs || [];
  const whitelist = options.whitelist || [];
  const whitelistRules = whitelist.map((rule) => new RegExp(rule));

  if (!enableServiceWorkers) {
    puppeteerArgs.push('--enable-features=NetworkService');
  }
  const browser =
    options.browser ||
    (await puppeteer.launch({
      args: puppeteerArgs,
    }));

  // All of these get mutated by the processPage() function. Once
  // per URL.
  const stylesheetAsts = {};
  const stylesheetContents = {};
  const doms = [];
  const allHrefs = [];
  const redirectResponses = {};
  const skippedUrls = new Set();

  try {
    for (const pageUrl of urls) {
      const page = await browser.newPage();
      if (!enableServiceWorkers) {
        await page._client.send('ServiceWorker.disable');
      }
      try {
        await processPage({
          page,
          htmlContent,
          headers,
          options,
          pageUrl,
          stylesheetAsts,
          stylesheetContents,
          doms,
          allHrefs,
          redirectResponses,
          skippedUrls,
        });
      } catch (e) {
        throw e;
      } finally {
        await page.close();
      }
    }
  } catch (e) {
    throw e;
  } finally {
    // We can close the browser now that all URLs have been opened.
    if (!options.browser) {
      browser.close();
    }
  }

  // All URLs have been opened, and we now have multiple DOM (cheerio) objects.
  // But first check that every spotted stylesheet (by <link> tags)
  // got downloaded.
  const missingASTs = [...allHrefs].filter((url) => {
    return !(
      stylesheetAsts[url] ||
      skippedUrls.has(url) ||
      redirectResponses[url]
    );
  });
  if (missingASTs.length) {
    throw new Error(
      `Found stylesheets that failed to download (${missingASTs})`
    );
  }

  // This is the protection against ever looking up the same CSS selector
  // more than once. We can confidently pre-populate it with a couple that
  // we're confident about.
  // The `''` selector is odd looking but comes from `:before` which is a
  // valid CSS selector.
  // See https://codepen.io/peterbe/pen/YBLyOd and
  // https://github.com/Semantic-Org/Semantic-UI/blob/master/dist/components/reset.min.css
  const decisionsCache = { '*': true, body: true, html: true, '': true };

  // Now, let's loop over ALL links and process their ASTs compared to
  // the DOMs.
  const isSelectorMatchToAnyElement = (selectorString) => {
    if (whitelistRules.some((regex) => regex.test(selectorString))) {
      return true;
    }
    // Here's the crucial part. Decide whether to keep the selector
    // Find at least 1 DOM that contains an object that matches
    // this selector string.
    return doms.some((dom) => {
      try {
        return dom(selectorString).length > 0;
      } catch (ex) {
        // Be conservative. If we can't understand the selector,
        // best to leave it in.
        if (debug) {
          console.warn(selectorString, ex.toString());
        }
        return true;
      }
    });
  };
  allHrefs.forEach((href) => {
    while (redirectResponses[href]) {
      href = redirectResponses[href];
    }
    if (skippedUrls.has(href)) {
      // skippedUrls are URLs that for some reason was deliberately not
      // downloaded. You can supply a `options.skippable` function which
      // might, for some reason, skip certain URLs. But if we don't
      // remember which URLs we skipped, when we later find all the
      // <link> tags to start analyze, we'd get an error here because
      // we deliberately chose to now parse its CSS.
      return;
    }
    const ast = stylesheetAsts[href];
    csstree.walk(ast, {
      visit: 'Rule',
      enter: function (node, item, list) {
        if (
          this.atrule &&
          csstree.keyword(this.atrule.name).basename === 'keyframes'
        ) {
          // Don't bother inspecting rules that are inside a keyframe.
          return;
        }

        if (!node.prelude.children) {
          const cssErrorMessage = `Invalid CSS found while evaluating ${href}: "${node.prelude.value}"`;
          if (options.ignoreCSSErrors) {
            list.remove(item);
          } else {
            throw new Error(cssErrorMessage);
          }
        } else {
          node.prelude.children.forEach((node, item, list) => {
            // Translate selector's AST to a string and filter pseudos from it
            // This changes things like `a.button:active` to `a.button`
            const selectorString = utils.reduceCSSSelector(
              csstree.generate(node)
            );

            // Before we begin, do a little warmup of the decision cache.
            // From a given selector, e.g. `div.foo p.bar`, we can first look
            // up if there's an point by first doing a lookup for `div.foo`
            // because if that doesn't exist we *know*  we can ignore more
            // "deeper" selectors like `div.foo p.bar` and `div.foo span a`.
            const parentSelectors = utils.getParentSelectors(selectorString);

            // If "selectorString" was `.foo .bar span`, then
            // this `parentSelectors` array will be
            // `['.foo', '.foo .bar']`.
            // If `selectorString` was just `.foo`, then
            // this `parentSelectors` array will be `[]`.
            let bother = true;
            parentSelectors.forEach((selectorParentString) => {
              if (bother) {
                // Is it NOT in the decision cache?
                if (selectorParentString in decisionsCache === false) {
                  decisionsCache[
                    selectorParentString
                  ] = isSelectorMatchToAnyElement(selectorParentString);
                }
                // What was the outcome of that? And if the outcome was
                // that it was NOT there, set the 'bother' to false which
                // will popoulate the decision cache immediately.
                if (!decisionsCache[selectorParentString]) {
                  bother = false;
                  decisionsCache[selectorString] = false;
                }
              } else {
                decisionsCache[selectorParentString] = false;
              }
            });

            if (selectorString in decisionsCache === false) {
              decisionsCache[selectorString] = isSelectorMatchToAnyElement(
                selectorString
              );
            }
            if (!decisionsCache[selectorString]) {
              // delete selector from a list of selectors
              list.remove(item);
            }
          });

          if (node.prelude.children.isEmpty()) {
            // delete rule from a list
            list.remove(item);
          }
        }
      },
    });
  });
  // Every unique URL in every <link> tag has been checked.
  // We can't simply loop over allHrefs because it might contain
  // entries that *aren't* in stylesheetAsts. For example, a page
  // might have a href in there but it's been deliberately skipped.
  // That's why we need to make a ordered copy of it based on
  // each item existing in stylesheetAsts.
  const allUsedHrefs = [];
  allHrefs.forEach((href) => {
    while (redirectResponses[href]) {
      href = redirectResponses[href];
    }
    if (stylesheetAsts[href]) {
      allUsedHrefs.push(href);
    }
  });
  const allCombinedAst = {
    type: 'StyleSheet',
    loc: null,
    children: allUsedHrefs.reduce(
      (children, href) => children.appendList(stylesheetAsts[href].children),
      new csstree.List()
    ),
  };

  // Lift important comments (i.e. /*! comment */) up to the beginning
  const comments = new csstree.List();
  csstree.walk(allCombinedAst, {
    visit: 'Comment',
    enter: (_node, item, list) => {
      comments.append(list.remove(item));
    },
  });
  allCombinedAst.children.prependList(comments);

  // Why not just allow the return of the "unminified" CSS (in case
  // some odd ball wants it)?
  // 'allCombinedAst' concatenates multiple payloads of CSS.
  // It only contains the selectors that are supposedly
  // in the DOM. However it does contain *duplicate* selectors.
  // E.g. `p { color: blue; } p { font-weight: bold; }`
  // When ultimately, what was need is `p { color: blue; font-weight: bold}`.
  // The csso.minify() function will solve this, *and* whitespace minify
  // it too.
  csso.syntax.compress(allCombinedAst, cssoOptions);
  postProcessOptimize(allCombinedAst);
  const finalCss = csstree.generate(allCombinedAst);
  const returned = {
    finalCss,
    stylesheetContents,
    doms,
  };
  return Promise.resolve(returned);
};

module.exports = { run: minimalcss };
